pkg/envfuncs/provider_funcs.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package envfuncs
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/e2e-framework/klient"
+	"sigs.k8s.io/e2e-framework/klient/k8s/resources"
+	"sigs.k8s.io/e2e-framework/klient/wait"
+	"sigs.k8s.io/e2e-framework/klient/wait/conditions"
+	"sigs.k8s.io/e2e-framework/pkg/env"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/support"
+	"sigs.k8s.io/e2e-framework/support/kind"
+)
+
+type clusterNameContextKey string
+
+// Following placeholder interface types are defined purely to make the pointer receiver work seemlessly
+// for the generic handlers. Without this, making the pointer receiver is really painful or we have to
+// turn providers into value receivers and that doesn't really work well for the current implementation
+// we have for kind provider. xref: https://t.ly/GDxOB
+
+// Every time you add a new provider that implements the support.E2EClusterProvider interface, that
+// needs to be listed here in this interface type with properly or'ing that with the existing
+// values.
+//
+//	type provider interfae {
+//			kind.Cluster | k3d.Cluster | kwok.Cluster
+//	}
+type provider interface {
```

Comment:
I am a bit confused why this interface extends a kind.Cluster type.
Is `type provider` kind specific?

Reply:
These hacked up two interface mechanism is the only way to make pointer recievers work with go generics right now. I have not found a much simpler solution :( 

xref: https://t.ly/GDxOB

Reply:
>  Is type provider kind specific?

Not it is not. That is why I also made sure to add comments. Say we add a new Provider like kwok,  then this line has to be updated to `kind.Cluster | kwok.Cluster`

Reply:
> These hacked up two interface mechanism is the only way to make pointer recievers work with go generics right now. I have not found a much simpler solution :(
> 
> xref: https://t.ly/GDxOB

@harshanarayana As the link points out, this Go approach is a bit clunky and is a workaround.
Until you pointed it out, I hadn't paid much attention to such issues with Go generics.